### DISCUSSION TITLE
Qt based build without Pixman, Pango, Cairo

### DISCUSSION THREAD

**Discussion by probonopd** [WayfireWM/wayfire#2017] [Ideas]
Hello, I am looking for a wlroots based Wayland compositor and window manager capable of rendering server side decorations for use in a [lightweight Qt based desktop environment](https://hellosystem.github.io/docs/).

Currently I am testing Raspberry Pi OS (which is using Wayfire), and am rather pleasantly surprised. (Things that never worked properly for me in other Wayland environments "just work" there, such as screen recording with sound.)

However, I would prefer not to have dependencies on the Gnome stack, and hence would like to see a variant of Wayfire that replaces the `pixman-dev cairo-dev pango-dev` dependencies with Qt.

I know that multiple Qt based lightweight desktops are also overwhelmed by the size of KWin and would probably like something like a QtWayfire compositor as well.

So,

- Would replacing Pixman, Pango, Cairo with Qt be doable or are they elementary to the way Wayfire works?
- Am I right in assuming that these are mainly used to draw the server side window decorations, and could be replaced by Qt for that?
- Would it require major architectural changes?
- Would this be something this project would be interested in having?

**Comment by ammen99** [comment]
Pango and cairo are used only by a few plugins which need to render text. Currently, this includes the decoration plugin, the scale plugin and the wsets plugin in wayfire 0.8.0. These can be easily replaced if you need to - we use cairo/pango to draw to an off-screen buffer, and then upload the data to OpenGL. It should be rather easy to replace the appropriate functionality with corresponding Qt calls.

Pixman is however central to Wayfire and even used inside wlroots, so getting rid of that is going to be hard. IMO it is not a big dependency so I do not really see why you'd want to remove it.

About upstreaming such changes: it might make sense to have a plugin 'text' that provides general text-rendering functionality similar to the helper we have now (cairo_text_t). The default implementation upstream would remain cairo-based, but you could then write a 'qt-text' plugin as an alternative to it. This way, plugins like scale and wsets will be usable without any patches. But anything more than this seems overly complicated (e.g. supporting decoration plugin with multiple backends).

**Reply by probonopd** [reply]
Agree, if Pixman is used by wlroots anyway, there is no need to remove it from Wayfire.
